openedx/core/djangoapps/programs/utils.py

```diff
@@ -127,6 +128,45 @@ def engaged_programs(self):
 
         return programs
 
+    def _is_course_in_progress(self, now, enrolled_run_modes, course):
+        """Check if course qualifies as in progress as part of the program.
+
+        A course is considered to be in progress if a user is enrolled in a run
+        of the correct mode or a run of the correct mode is still available for enrollment.
+
+        Arguments:
+            now (datetime): datetime for now
+            enrolled_run_modes (dict): Dict showing what mode is enrolled for each run.
+            course (dict): Containing nested course runs.
+
+        Returns:
+            bool, indicating whether the course is in progress.
+        """
+        # Part 1: Check if any of the seats you are enrolled in qualify this course as in progress
+        enrolled_runs = [run for run in course['course_runs'] if run['key'] in self.course_run_ids]
+        # Check if the user is enrolled in the required mode for the run
+        runs_with_required_mode = [
+            run for run in enrolled_runs
+            if run.get('type') == enrolled_run_modes[run['key']]
```

Comment:
i did this because a very large amount of tests failed on not having a type because it is not defined in the factory, but adding it to the factory broke some other tests in vary mysterious ways.  
For the tests we could add a type to every course run, a course run with a type to every course, and a course with a course run with a type to every program, but that will be very verbose, is it worth it if we can feel pretty confident that the catalog will always return this key?